File: src/converter/plugins/AnonymousDeclarationPlugin.ts
import ts, {Node} from "typescript";
import {ConverterPlugin} from "../plugin.js";
import {ConverterContext} from "../context.js";
import {Render} from "../render.js";
import {ConfigurationService, configurationServiceKey} from "./ConfigurationPlugin.js";
import {NameResolverService, nameResolverServiceKey} from "./NameResolverPlugin.js";
import {NamespaceInfoService, namespaceInfoServiceKey} from "./NamespaceInfoPlugin.js";
import {generateDerivedDeclarations} from "../../structure/derived/generateDerivedDeclarations.js";
import {DerivedDeclaration} from "../../structure/derived/derivedDeclaration.js";
import {TypeScriptService, typeScriptServiceKey} from "./TypeScriptPlugin.js";
import {DerivedFile} from "../generated.js";

export interface AnonymousDeclarationContext extends ConverterContext {
    resolveName(node: Node): string
}

export interface AnonymousDeclarationRenderResult {
    name: string
    declaration: string
    reference: string
}

export type AnonymousDeclarationRender = (
    node: ts.Node,
    context: AnonymousDeclarationContext,
    render: Render,
) => AnonymousDeclarationRenderResult | string | null

class AnonymousDeclarationPlugin<TNode extends Node = Node> implements ConverterPlugin {
    private readonly generated: DerivedDeclaration[] = []

    constructor(
        private anonymousDeclarationRender: AnonymousDeclarationRender,
    ) {
    }

    generate(context: ConverterContext): DerivedFile[] {
        const configurationService = context.lookupService<ConfigurationService>(configurationServiceKey)
        const configuration = configurationService?.configuration
        if (configuration === undefined) throw new Error("AnonymousDeclarationPlugin can't work without ConfigurationService")

        const namespaceInfoService = context.lookupService<NamespaceInfoService>(namespaceInfoServiceKey)
        const resolveNamespaceStrategy = namespaceInfoService?.resolveNamespaceStrategy?.bind(namespaceInfoService)
        if (resolveNamespaceStrategy === undefined) throw new Error("AnonymousDeclarationPlugin can't work without NamespaceInfoService")

        return generateDerivedDeclarations(this.generated, configuration, resolveNamespaceStrategy)
    }

    render(node: ts.Node, context: ConverterContext, next: Render): string | null {
        const nameResolverService = context.lookupService<NameResolverService>(nameResolverServiceKey)
        if (nameResolverService === undefined) throw new Error("AnonymousDeclarationPlugin can't work without NameResolverService")

        const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
        if (typeScriptService === undefined) throw new Error("AnonymousDeclarationPlugin can't work without TypeScriptService")

        const resolveName = (node: TNode) => nameResolverService.resolveName(node, context)

        const anonymousDeclarationContext = {
            ...context,
            resolveName,
        }

        const result = this.anonymousDeclarationRender(
            node,
            anonymousDeclarationContext,
            next
        )

        if (result === null || typeof result === "string") return result

        const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"
        const namespace = typeScriptService.findClosest(node, ts.isModuleDeclaration)

        const {name, declaration, reference} = result

        this.generated.push({
            sourceFileName,
            namespace,
            fileName: `${name}.kt`,
            body: declaration,
        })

        return reference;
    }

    setup(context: ConverterContext): void {
    }

    traverse(node: ts.Node, context: ConverterContext): void {
    }
}

export function createAnonymousDeclarationPlugin<TNode extends Node = Node>(
    render: AnonymousDeclarationRender,
): ConverterPlugin {
    return new AnonymousDeclarationPlugin<TNode>(render)
}

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/AnonymousDeclarationPlugin.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.*
import io.github.sgrishchenko.karakum.structure.derived.DerivedDeclaration
import io.github.sgrishchenko.karakum.structure.derived.generateDerivedDeclarations
import io.github.sgrishchenko.karakum.util.currentAbortable
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import js.array.ReadonlyArray
import js.promise.Promise
import js.promise.await
import kotlinx.js.JsPlainObject
import typescript.Node
import web.abort.Abortable

external interface AnonymousDeclaration

@JsPlainObject
internal external interface AnonymousDeclarationRenderResult {
    val name: String
    val declaration: String
    val reference: String
}

@Suppress("NOTHING_TO_INLINE")
inline fun AnonymousDeclaration(result: String) =
    result.unsafeCast<AnonymousDeclaration>()

@Suppress("NOTHING_TO_INLINE")
inline fun AnonymousDeclaration(
    name: String,
    declaration: String,
    reference: String,
) =
    AnonymousDeclarationRenderResult(
        name = name,
        declaration = declaration,
        reference = reference,
    ).unsafeCast<AnonymousDeclaration>()

typealias AnonymousDeclarationRender = suspend (
    node: Node,
    context: Context,
    render: Render<Node>,
) -> AnonymousDeclaration?

class AnonymousDeclarationPlugin(
    render: AnonymousDeclarationRender,
) : Plugin {
    private val anonymousDeclarationRender = render
    private val generated = mutableMapOf<Node, DerivedDeclaration>()

    override suspend fun setup(context: Context) = Unit

    override suspend fun traverse(node: Node, context: Context) = Unit

    override suspend fun render(node: Node, context: Context, next: Render<Node>): String? {
        val typeScriptService = context.requireService(typeScriptServiceKey)
        val annotationService = context.requireService(annotationServiceKey)

        val annotations = annotationService.resolveAnonymousAnnotations(node, context)

        val result = anonymousDeclarationRender(node, context, next)

        if (result == null) return result
        if (jsTypeOf(result) == "string") return result.toString()

        val renderResult = result.unsafeCast<AnonymousDeclarationRenderResult>()
        val name = renderResult.name
        val declaration = renderResult.declaration
        val reference = renderResult.reference

        val sourceFileName = node.getSourceFileOrNull()?.fileName ?: "generated.d.ts"
        val namespace = typeScriptService.findClosestNamespace(node)

        generated[node] = DerivedDeclaration(
            sourceFileName = sourceFileName,
            namespace = namespace,
            fileName = "${name}.kt",
            body = (annotations + declaration).joinToString(separator = "\n"),
        )

        return reference
    }

    override suspend fun generate(context: Context, render: Render<Node>): ReadonlyArray<DerivedFile> {
        return generateDerivedDeclarations(generated.values.toTypedArray(), context)
    }
}

fun createAnonymousDeclarationPlugin(
    render: AnonymousDeclarationRender,
): Plugin {
    return AnonymousDeclarationPlugin(render)
}

@JsExport
@JsName("createAnonymousDeclarationPlugin")
fun createAnonymousDeclarationPluginAsync(
    render: (
        node: Node,
        context: Context,
        render: Render<Node>,
        options: Abortable,
    ) -> Promise<AnonymousDeclaration?>,
): JsPlugin =
    AnonymousDeclarationPlugin { node, context, next ->
        render(node, context, next, currentAbortable()).await()
    }.toJsPlugin()
